protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,739 @@
+                            Some(PeersIterState::Waiting(None)) => {}
+                            Some(PeersIterState::WaitingAtCapacity) => {}
+                            Some(PeersIterState::Finished) => {
+                                // `state` is never set to `Finished`.
+                                unreachable!();
+                            },
+                            None => state = Some(PeersIterState::WaitingAtCapacity),
+                        };
+
+                        break;
+                    }
+                    PeersIterState::Finished => break,
+                }
+            }
+        }
+
+        state.unwrap_or(PeersIterState::Finished)
+    }
+
+    pub fn finish(&mut self) {
+        for iter in &mut self.iters {
+            iter.finish()
+        }
+    }
+
+    pub fn into_result(self) -> impl Iterator<Item = PeerId> {
+        let mut result = HashSet::new();
+
+        let mut iters = self.iters.into_iter().map(ClosestPeersIter::into_result).collect::<Vec<_>>();
+
+        'outer: loop {
+            let mut progress = false;
+
+            for iter in iters.iter_mut() {
+                if let Some(peer) = iter.next() {
+                    progress = true;
+                    result.insert(peer);
+                    if result.len() == self.config.num_results.get() {
+                        break 'outer;
+                    }
+                }
+            }
+
+            if !progress {
+                break;
+            }
+        }
+
+        result.into_iter()
+    }
+}
+
+/// Index into the [`ClosestDisjointPeersIter`] `iters` vector.
+#[derive(Copy, Clone)]
+struct IteratorIndex(usize);
+
+impl From<usize> for IteratorIndex {
+    fn from(i: usize) -> Self {
+        IteratorIndex(i)
+    }
+}
+
+impl From<IteratorIndex> for usize {
+    fn from(i: IteratorIndex) -> Self {
+        i.0
+    }
+}
+
+impl Add<usize> for IteratorIndex {
+    type Output = Self;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        (self.0 + rhs).into()
+    }
+}
+
+impl Index<IteratorIndex> for Vec<ClosestPeersIter> {
+    type Output = ClosestPeersIter;
+
+    fn index(&self, index: IteratorIndex) -> &Self::Output {
+        &self[index.0]
+    }
+}
+
+impl IndexMut<IteratorIndex> for Vec<ClosestPeersIter> {
+    fn index_mut(&mut self, index: IteratorIndex) -> &mut Self::Output {
+        &mut self[index.0]
+    }
+}
+
+/// State tracking the iterators that yielded (i.e. tried to contact) a peer. See
+/// [`ClosestDisjointPeersIter::on_success`] for details.
+struct PeerState {
+    /// First iterator to yield the peer. Will be notified both of the outcome
+    /// (success/failure) as well as the closer peers.
+    initiated_by: IteratorIndex,
+    /// Additional iterators only notified of the outcome (success/failure), not
+    /// the closer peers, in order to uphold the S/Kademlia disjoint paths
+    /// guarantee.
+    additionally_awaited_by: Vec<IteratorIndex>,
```

Comment:
As mentioned elsewhere, I think we don't need this (but am happy to be corrected).